File: 프로그래머스/2/258711. 도넛과 막대 그래프/도넛과 막대 그래프.py
from collections import defaultdict, deque

def dfs(graph, start, visited):
    node_count, edge_count = 0, 0
    stack = [start]
    
    while stack:
        current = stack.pop()
        if visited[current]:
            continue
        
        visited[current] = True
        node_count += 1
        
        for i in graph[current]:
            edge_count += 1
            if not visited[i]:
                stack.append(i)

    return edge_count, node_count

def solution(edges):
    graph = defaultdict(list)
    in_degree = defaultdict(int)

    max_node = 0
    for a, b in edges:
        graph[a].append(b)
        in_degree[b] += 1
        max_node = max(max_node, a, b)

    visited = [False] * (max_node + 1)

    # 생성한 정점의 번호 찾기
    new_node = None
    for i in range(1, max_node + 1):
        if in_degree[i] == 0 and len(graph[i]) > 1:
            new_node = i
            break

    answer = [new_node, 0, 0, 0]
    
    # 그래프 모양 확인하기
    for i in graph[new_node]:
        if visited[i]:
            continue

        edge_count, node_count = dfs(graph, i, visited)

        if edge_count == node_count:
            answer[1] += 1
        elif edge_count + 1 == node_count:
            answer[2] += 1
        else:
            answer[3] += 1

    return answer

File: 프로그래머스/2/258711. 도넛과 막대 그래프/도넛과 막대 그래프.java
import java.util.*;

class Solution {
    public int[] solution(int[][] edges) {
        int[] answer = new int[4];
        
        Map<Integer, List<Integer>> graph = new HashMap<>();
        int[] inDegree = new int[1000001];
        int[] outDegree = new int[1000001];
        
        for (int[] edge : edges) {
            graph.computeIfAbsent(edge[0], k -> new ArrayList<>()).add(edge[1]);
            inDegree[edge[1]]++;
            outDegree[edge[0]]++;
        }
        
        for (int i = 0; i < outDegree.length; i++) {
            if (outDegree[i+1] >= 2 && inDegree[i+1] == 0) {
                answer[0] = i+1;
                break;
            }
        }
        
        boolean[] visited = new boolean[1000001];
        for (int i : graph.get(answer[0])) {
            if (visited[i]) continue;
            
            int idx = dfs(graph, i, visited);
            answer[idx]++;
        }
        return answer;
    }
    
    public int dfs(Map<Integer, List<Integer>> graph, int start, boolean[] visited) {
        Stack<Integer> stack = new Stack<>();
        stack.push(start);
        
        int nodeCount = 0, edgeCount = 0;
        while (!stack.isEmpty()) {
            int now = stack.pop();
            if (visited[now]) continue;
            
            visited[now] = true;
            nodeCount += 1;
            
            for (int next : graph.getOrDefault(now, new ArrayList<>())) {
                edgeCount += 1;
                if (!visited[next]) {
                    stack.push(next);
                }
            }
        }
        
        if (nodeCount == edgeCount) {
            return 1;
        } else if (nodeCount == edgeCount + 1) {
            return 2;
        } else {
            return 3;
        }
    }
}